src/util/bson/bson.cpp

```diff
@@ -443,641 +444,196 @@ std::ostream& operator<<(std::ostream& out, const Bson& b)
-            if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
-                // we will ignore the subtype for now
-                // pop vestigal binary instruction
-                m_instructions.pop();
-                m_instructions.push({State::Skip});
-                m_instructions.push({State::Skip});
-            } else {
-                // we will ignore the subtype for now
-                m_marks.top().push_back(instruction.key, std::vector<char>());
-            }
-
-            break;
-        case State::BinaryBase64: {
-            // if we have already pushed a binary type
-            if (m_marks.top().size() && m_marks.top().back().first == instruction.key) {
-                m_marks.top().pop_back();
-                m_marks.top().push_back(instruction.key, std::vector<char>(val.begin(), val.end()));
-
-                // pop vestigal binary instruction
-                m_instructions.pop();
-                m_instructions.push({State::Skip});
-                m_instructions.push({State::Skip});
-            } else {
-                // we will ignore the subtype for now
-                m_marks.top().push_back(instruction.key, std::vector<char>(val.begin(), val.end()));
-            }
-
-            break;
+            return Bson(out);
         }
-        default:
-            m_marks.top().push_back(instruction.key, std::string(val.begin(), val.end()));
-            break;
+        case Json::value_t::discarded:
+            REALM_ASSERT_RELEASE(!"should never see discarded");
     }
-    return true;
-}
-
-/*!
- @brief an object key was read
- @param[in] val  object key
- @return whether parsing should proceed
- @note It is safe to move the passed string.
- */
-bool Parser::key(string_t& val) {
-    if (!m_instructions.empty()) {
-        auto top = m_instructions.top();
-
-        if (top.type == State::RegularExpression) {
-            if (val == key_regular_expression_pattern) {
-                m_instructions.push({State::RegularExpressionPattern, top.key});
-            } else if (val == key_regular_expression_options) {
-                m_instructions.push({State::RegularExpressionOptions, top.key});
-            }
-            return true;
-        } else if (top.type == State::Date) {
-            return true;
-        } else if (top.type == State::Binary) {
-            if (val == key_binary_base64) {
-                m_instructions.push({State::BinaryBase64, top.key});
-            } else if (val == key_binary_sub_type) {
-                m_instructions.push({State::BinarySubType, top.key});
+    REALM_ASSERT_RELEASE(!"unknown json value type");
+}
+
+// This works around the deleted rvalue constructor in StringData
+inline StringData tosd(const std::string& s) { return s; }
+
+// Keep these sorted by key. This is checked so you can't forget.
+using FancyParser = Bson(*)(const Json& json);
+static constexpr std::pair<std::string_view, FancyParser> bson_fancy_parsers[] = {
```

Comment:
Using a sorted array rather than a map means that this is part of the binary image rather than being constructed at startup.

Reply:
👍